### PR TITLE
Add predict step accumulation

### LINKED ISSUE #7232: trainer.evaluate() aggregates predictions on GPU and causes CUDA out of memory issues for large datasets
## Environment info
- `transformers` version: 3.1.0
- Platform: Linux-4.15.0-112-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.6.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help
 
Trainer: @sgugger

## Information

Model I am using (Bert, XLNet ...): Bert

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

## To reproduce

1. Use the Trainer for evaluation (`.evaluate()`, `.predict()`) on the GPU with BERT with a large evaluation DataSet where the size of the returned prediction Tensors + Model exceed GPU RAM. (In my case I had an evaluation dataset of 469,530 sentences).
2. Trainer will crash with a CUDA Memory Exception

## Expected behavior

- I would expect the predictions in `predict()` or `evaluate()` from each step would be moved to the CPU device (off the GPU) and then concatenated later. However, the tensors are concatenated while still be on the GPU device, and only converted to CPU numpy arrays after the whole dataset has been predicted/evaluated. 

This means that for large evaluation datasets you'll run out of CUDA memory.

It also makes it difficult to pick the batch size to optimize the batch size for the GPU, as you need to allow space for not only the model and inputs, but also all the predictions, which can add up when dealing with large evaluation datasets.

- ie. The problem in the trainer code is that the predictions stay on the GPU [here](https://github.com/huggingface/transformers/blob/67d9fc50d917c63cf67281106214e1d9ae018dff/src/transformers/trainer.py#L1315)
- These tensors get concatenated but stay on the GPU [here](https://github.com/huggingface/transformers/blob/67d9fc50d917c63cf67281106214e1d9ae018dff/src/transformers/trainer_utils.py#L143)
- and then the predictions eventually end up on the CPU [here](https://github.com/huggingface/transformers/blob/67d9fc50d917c63cf67281106214e1d9ae018dff/src/transformers/trainer.py#L1343-L1347)

What this means, is that for larger evaluation datasets, *all* the predictions stay on the GPU, which is a function of how long your evaluation dataset is, and that you often run out of GPU RAM.

A work around is to something like this and run the loop yourself, and predict by batch:

``` py
    preds = []
    for i in tqdm(range(0, len(ds_valid), step)):
        ds_valid = KeywordDataset(tokenizer, df_valid[i:i+step], targets)
        batch_preds = trainer.predict(ds_valid)
        preds.append(batch_preds)
        batch_accuracy = (batch_preds.predictions.argmax(-1) == batch_preds.label_ids).mean()

    np_preds = np.concatenate([pred.predictions for pred in preds], axis=0)
    np_label_ids = np.concatenate([pred.label_ids for pred in preds], axis=0)
    acc = (np_preds.argmax(-1) == np_label_ids).mean()
    print('eval_accuracy = ', acc)
``` 

It would be nice if it defaulted the cpu (so you don't have to worry about it), and that there is another trainer argument to give the device to aggregate predictions that can be used to override that behaviour if it's important for certain use cases.

Thanks for the amazing work on an amazing library. Working with transformers has never been easier due to the hard work of your team!

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

Currently, the Trainer accumulates all predictions on the host (GPU or TPU) before gathering across all hosts (in case of distributed training) and moving back to the CPU. This can result in OOM errors when users have a big dataset (the model is already taking up a lot of space on the host) as highlighted in #7232. However moving the predictions to the CPU at each prediction step is also inefficient (particularly on TPU).

This PR aims at fixing the OOM problem while retaining efficiency by introducing a new training argument called `eval_accumulation_step`. If left untouched, the behavior is the same as right now (all predictions accumulated on the host and moved at the end of the prediction loop). If set to an int, the predictions are gathered and moved every `eval_accumulation_step`. This required some clever reorganization of the predictions (see the docstring of `DistributedTensorGatherer` for more details).

In passing I cleaned up the code related to gathering tensors across multiple hosts and fixed the issue of the `loss.item()` (big slow down to do that at every step on TPUs) and accumulated the losses the same way predictions and labels are. This still works for any number of outputs/labels of the model.

To check those changes did not break anything, I ran `test_trainer_distributed.py` on my local setup and created an equivalent for TPUs that I also ran (they both pass).

This slightly change Seq2SeqTrainer (since we don't want the `loss.item()`) so cc @patil-suraj I don't think this should break anything in it.

Fixes #7232

### CODE REVIEW

**Review by julien-c — APPROVED** [review]
Wow, this is really cool! Great job! 

For the TPU perf script, should we add that to our TPU CI to make sure perf doesn’t regress?

File: src/transformers/trainer_pt_utils.py [hunk 1]
```diff
@@ -177,3 +186,90 @@ def get_tpu_sampler(dataset: torch.utils.data.dataset.Dataset):
...
+        return type(tensors)(nested_truncate(t, limit) for t in tensors)
+    return tensors[:limit]
+
+
+class DistributedTensorGatherer:
+    """
+    A class responsible for properly gathering tensors (or nested list/tuple of tensors) on the CPU
+    by chunks.
+
+    If our dataset as 15 samples with a batch size of 2 on 3 processes and we gather then transfer on
```
Comment (LysandreJik) [inline src/transformers/trainer_pt_utils.py:210]:
```suggestion
    If our dataset has 15 samples with a batch size of 2 on 3 processes and we gather then transfer on
```

File: src/transformers/trainer_pt_utils.py [hunk 2]
```diff
@@ -177,3 +186,90 @@ def get_tpu_sampler(dataset: torch.utils.data.dataset.Dataset):
...
+    CPU at every step, our sampler will generate the following indices:
+
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 0, 1, 2]
```
Comment (LysandreJik) [inline src/transformers/trainer_pt_utils.py:213]: Aren't we computing the results twice on samples 0, 1 and 2?

**Review by LysandreJik — APPROVED** [review]
Very nice, way cleaner!

You can add your TPU tests to [test_xla_examples.py](https://github.com/huggingface/transformers/blob/master/examples/test_xla_examples.py) for now and lets refactor later to have a better suite for TPU.

File: src/transformers/trainer_pt_utils.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/trainer_pt_utils.py:213]: Yes we are, the distributed sampler used makes sure all processes get the same length of dataset by circling through the indices. Though for my example 15 is a multiple of 3 so I should have set 16 elements...
That's why `finalize` has a truncation operation, to forget those elements we added.

Comment (LysandreJik) [reply src/transformers/trainer_pt_utils.py:213]: Great, thanks for the explanation.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @sgugger

**Comment by eugeneware** [comment]
Thanks so much @sgugger - you're a legend! 🙇

**Comment by Godwinh19** [comment]
> # What does this PR do?
> Currently, the Trainer accumulates all predictions on the host (GPU or TPU) before gathering across all hosts (in case of distributed training) and moving back to the CPU. This can result in OOM errors when users have a big dataset (the model is already taking up a lot of space on the host) as highlighted in #7232. However moving the predictions to the CPU at each prediction step is also inefficient (particularly on TPU).
> 
> This PR aims at fixing the OOM problem while retaining efficiency by introducing a new training argument called `eval_accumulation_step`. If left untouched, the behavior is the same as right now (all predictions accumulated on the host and moved at the end of the prediction loop). If set to an int, the predictions are gathered and moved every `eval_accumulation_step`. This required some clever reorganization of the predictions (see the docstring of `DistributedTensorGatherer` for more details).
> 
> In passing I cleaned up the code related to gathering tensors across multiple hosts and fixed the issue of the `loss.item()` (big slow down to do that at every step on TPUs) and accumulated the losses the same way predictions and labels are. This still works for any number of outputs/labels of the model.
> 
> To check those changes did not break anything, I ran `test_trainer_distributed.py` on my local setup and created an equivalent for TPUs that I also ran (they both pass).
> 
> This slightly change Seq2SeqTrainer (since we don't want the `loss.item()`) so cc @patil-suraj I don't think this should break anything in it.
> 
> Fixes #7232

`eval_accumulation_steps` for the argument name and not `eval_accumulation_step` 😉

**Pull request merged**